README.md
# vs_scrollbar  [![Pub](https://img.shields.io/pub/v/vs_scrollbar.svg)](https://pub.dartlang.org/packages/vs_scrollbar)

Customizable ScrollBar that can be dragged for quick navigation supporting both Horizontal and Vertical list.

---

## Table of Contents

- [Installing](#installing) - How to install
- [Package Details](package-details) - Package Details
- [Example](#example) - How this plugin works
- [Code example](#code-example) - How to use
- [VsScrollBar object properties](#vsscrollbar-object-properties) - Modifying your VsScrollBar

---

## Installing

Add to pubspec.yaml file

```sh
dependencies:
  vs_scrollbar: ^1.0.0
```

Import

```sh
import 'package:vs_scrollbar/vs_scrollbar.dart';
```

## Package Details

Customizable ScrollBar that can be dragged for quick navigation supporting both Horizontal and Vertical list.

You can customize scrollbar's ```thickness```, ```radius```, set custom ```colors``` and handle support for list drag functionality with prebuilt methods of default ScrollBar for Flutter.

Control animation duration of scrollbar using scrollbarTimeToFade, scrollbarFadeDuration properties.

Drag Behaviour with scrollbar is by default supported, if ScrollController object is passed.

Wrap This ScrollBar to any ScrollableList.  
```SingleChildScrollView```  
```GridView```  
```ListView```  
```Listview.builder```

## Example 
Try out example code all parameters are optional, currently serving basic use case scenarios.

## Horizontal List

![](https://github.com/VickySalunkhe/vs_scrollbar/blob/main/example/assets/horizontal.webp)

---

## Vertical List
![](https://github.com/VickySalunkhe/vs_scrollbar/blob/main/example/assets/vertical.webp)

### Code example

```dart
// use same scrollController object to listview and scrollbar to support drag functionality
ScrollController _scrollController = ScrollController();

VsScrollbar(
   controller: _scrollController,
   showTrackOnHover: true,// default false
   isAlwaysShown: true, // default false
   scrollbarFadeDuration: Duration(milliseconds: 500), // default : Duration(milliseconds: 300)
   scrollbarTimeToFade: Duration(milliseconds: 800),// default : Duration(milliseconds: 600)
   style: VsScrollbarStyle(
            hoverThickness: 10.0, // default 12.0
            radius: Radius.circular(10), // default Radius.circular(8.0)
            thickness: 10.0, // [ default 8.0 ]
            color: Colors.purple.shade900, // default ColorScheme Theme
          ),
   child: ListView.builder(
             controller: _scrollController, 
             shrinkWrap: true,
             physics: BouncingScrollPhysics(),
             itemCount: 6,
             scrollDirection: Axis.vertical,
             itemBuilder: (BuildContext context, int index) {
                   return Container(
                          height: 100,
                          width: MediaQuery.of(context).size.width,
                          color: Colors.purple[900],
                          margin: EdgeInsets.all(15));
             },
   ),
)
```
---


## VsScrollbar object properties

| Name                    | Type                | default                               | Description                                                                  |
| ----------------------- | ------------------- | ------------------------------------- | ---------------------------------------------------------------------------- |
| controller              | `ScrollController`  | null                                  | Controller for Scrollbar                                                     |
| scrollbarFadeDuration   | `Duration`          | Duration(milliseconds: 300)           | Sets scrollbar fade animation duration                                       |
| scrollbarTimeToFade     | `Duration`          | Duration(milliseconds: 600)           | Fades scrollbar after certain duration                                       |
| isAlwaysShown           | `bool`              | false                                 | controls visibility of scrollbar when ideal                                  |
| showTrackOnHover        | `bool`              | false                                 | controls visibility of scrollbar while hovering                              |
| style                   | `VsScrollbarStyle`  | VsScrollbarStyle()                    | Style Class for VsScrollBar                                                  |


## VsScrollbarStyle object properties

| Name                    | Type                | default                               | Description                                                                  |
| ----------------------- | ------------------- | ------------------------------------- | ---------------------------------------------------------------------------- |
| color                   | `Color`             | System ColorScheme Theme              | ScrollBar Color                                                              |
| radius                  | `Radius`            | Radius.circular(8.0)                  | Set Radius of ScrollBar                                                      |
| thickness               | `double`            | 8.0                                   | Set Thickness of ScrollBar                                                   |
| hoverThickness          | `double`            | 12.0                                  | Set Hover Thickness of ScrollBar                                             |

---

## Pull request and feedback are always appreciated :)

---

### Show some :heart: and star the repo.

[![GitHub followers](https://img.shields.io/github/followers/VickySalunkhe.svg?style=social&label=Follow)](https://github.com/VickySalunkhe)
[![Twitter Follow](https://img.shields.io/twitter/follow/VickySalunkhe.svg?style=social)](https://twitter.com/vickysalunkhe01)

---

### :heart: Found this project useful?

If you found this project useful, then please consider giving it a :star: on Github and sharing it with your friends via social media.

---

<a href="https://www.buymeacoffee.com/VickySalunkhe" target="_blank"><img src="https://img.buymeacoffee.com/button-api/?text=Buy me a coffee&emoji=&slug=VickySalunkhe&button_colour=5F7FFF&font_colour=ffffff&font_family=Cookie&outline_colour=000000&coffee_colour=FFDD00"></a>

---




CHANGELOG.md
## [1.0.0] - 15/09/2023
* Updated Supported Dart SDK Version  sdk: ">=2.12.0 <4.0.0"
* Replaced Depreciated Properties with new properties.
* Changed Example Repo

## [0.2.1] - 27/06/2021
*Issue Error: isPointerOverScrollbar Fixed
## [0.2.0] - 16/03/2021

*Breaking Changes
*Package Migrated to Dart Sound Null Safety
*Added VsScrollbarStyle Class for Scrollbar Style
*Added Drag support by default, works, if ScrollController object is passed.
*Added hover property controllers thickness and visibility

## [0.1.2] - 15/01/2021

*Added use case for preventing scrolling of scrollbar thorugh horizontal drag on vertical scrolling list and vice versa.

## [0.1.1] - 13/01/2021

*Added scrollbarTimeToFade & scrollbarFadeDuration fields using which you can control duration of animation of scrollbar.

## [0.1.0] - 28/12/2020

*Added List Drag Support for both horizontal vertical list.

## [0.0.2] - 25/12/2020

*Minor Changes

## [0.0.1] - 16/12/2020

 *Initial release.

lib/vs_scrollbar.dart
library vs_scrollbar;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

class VsScrollbarStyle {
  /// The hoverThickness of the VsScrollbar thumb.
  /// default value is 12.0 pixels.
  final double hoverThickness;

  /// The thickness of the VsScrollbar thumb.
  /// default thickness of 8.0 pixels.
  final double thickness;

  /// The radius of the VsScrollbar thumb.
  /// default [Radius.circular] of 8.0 pixels.
  final Radius radius;

  /// The color of the VsScrollbar thumb.
  final Color? color;

  const VsScrollbarStyle(
      {this.radius = _kScrollbarRadius,
      this.thickness = _kScrollbarThickness,
      this.hoverThickness = _kScrollbarThicknessWithTrack,
      this.color});
}

const VsScrollbarStyle _kScrollbarStyle = const VsScrollbarStyle();
const double _kScrollbarThickness = 8.0;
const double _kScrollbarThicknessWithTrack = 12.0;
const double _kScrollbarMargin = 2.0;
const double _kScrollbarMinLength = 48.0;
const Radius _kScrollbarRadius = Radius.circular(8.0);
const Duration _kScrollbarFadeDuration = Duration(milliseconds: 300);
const Duration _kScrollbarTimeToFade = Duration(milliseconds: 600);

/// To add a VsScrollbar to a [ScrollView], wrap the scroll view
/// widget in a [VsScrollbar] widget.
///
/// The color of the VsScrollbar will change when dragged. A hover animation is
/// also triggered when used on web and desktop platforms. A VsScrollbar track
/// can also been drawn when triggered by a hover event, which is controlled by
/// [showTrackOnHover]. The thickness of the track and VsScrollbar thumb will
/// become larger when hovering, unless overridden by [hoverThickness].
///
/// See also:
///
///  * [RawScrollbar], a basic VsScrollbar that fades in and out, extended
///    by this class to add more animations and behaviors.
///  * [ScrollbarTheme], which configures the VsScrollbar's appearance.
///  * [ListView], which displays a linear, scrollable list of children.
///  * [GridView], which displays a 2 dimensional, scrollable array of children.
class VsScrollbar extends StatefulWidget {
  /// Creates a material design VsScrollbar that by default will connect to the
  /// closest Scrollable descendant of [child].
  ///
  /// The [child] should be a source of [ScrollNotification] notifications,
  /// typically a [Scrollable] widget.
  ///
  /// If the [controller] is null, the default behavior is to
  /// enable VsScrollbar dragging using the [PrimaryScrollController].
  ///
  const VsScrollbar({
    Key? key,
    required this.child,
    this.controller,
    this.style = _kScrollbarStyle,
    this.scrollbarFadeDuration,
    this.scrollbarTimeToFade,
    this.isAlwaysShown,
    this.showTrackOnHover,
    this.notificationPredicate,
  }) : super(key: key);

  /// {@macro flutter.widgets.VsScrollbar.child}
  final Widget child;

  /// {@macro flutter.widgets.VsScrollbar.controller}
  final ScrollController? controller;

  /// {@macro flutter.widgets.VsScrollbar.isAlwaysShown}
  final bool? isAlwaysShown;

  /// If this property is null, then [ScrollbarThemeData.trackVisibility] of
  /// [ThemeData.scrollbarTheme] is used. If that is also null, the default value
  /// is false.
  final bool? showTrackOnHover;

  ///Style Property for VsScrollbar
  final VsScrollbarStyle style;

  /// {@macro flutter.widgets.VsScrollbar.notificationPredicate}
  final ScrollNotificationPredicate? notificationPredicate;

  /// default 600ms
  final Duration? scrollbarTimeToFade;

  /// default 300ms
  final Duration? scrollbarFadeDuration;

  @override
  _ScrollbarState createState() => _ScrollbarState();
}

class _ScrollbarState extends State<VsScrollbar> {
  @override
  Widget build(BuildContext context) {
    return _MaterialScrollbar(
      child: widget.child,
      controller: widget.controller,
      isAlwaysShown: widget.isAlwaysShown,
      showTrackOnHover: widget.showTrackOnHover,
      hoverThickness: widget.style.hoverThickness,
      thickness: widget.style.thickness,
      radius: widget.style.radius,
      color: widget.style.color,
      notificationPredicate: widget.notificationPredicate,
    );
  }
}

class _MaterialScrollbar extends RawScrollbar {
  const _MaterialScrollbar({
    Key? key,
    required Widget child,
    ScrollController? controller,
    bool? isAlwaysShown,
    this.showTrackOnHover,
    this.hoverThickness,
    this.color,
    this.scrollbarFadeDuration,
    this.scrollbarTimeToFade,
    double? thickness,
    Radius? radius,
    ScrollNotificationPredicate? notificationPredicate,
  }) : super(
          key: key,
          child: child,
          controller: controller,
          thumbVisibility: isAlwaysShown,
          thickness: thickness,
          radius: radius,
          fadeDuration: scrollbarFadeDuration ?? _kScrollbarFadeDuration,
          timeToFade: scrollbarTimeToFade ?? _kScrollbarTimeToFade,
          pressDuration: Duration.zero,
          notificationPredicate:
              notificationPredicate ?? defaultScrollNotificationPredicate,
        );
  final Duration? scrollbarTimeToFade;
  final Duration? scrollbarFadeDuration;
  final Color? color;
  final bool? showTrackOnHover;
  final double? hoverThickness;

  @override
  _MaterialScrollbarState createState() => _MaterialScrollbarState();
}

class _MaterialScrollbarState extends RawScrollbarState<_MaterialScrollbar> {
  late AnimationController _hoverAnimationController;
  bool _dragIsActive = false;
  bool _hoverIsActive = false;
  late ColorScheme _colorScheme;
  late ScrollbarThemeData _scrollbarTheme;
  // On Android, scrollbars should match native appearance.
  late bool _useAndroidScrollbar;

  @override
  bool get showScrollbar => widget.thumbVisibility ?? false;

  bool get _showTrackOnHover => widget.showTrackOnHover ?? false;

  Set<MaterialState> get _states => <MaterialState>{
        if (_dragIsActive) MaterialState.dragged,
        if (_hoverIsActive) MaterialState.hovered,
      };

  MaterialStateProperty<Color> get _thumbColor {
    final Color onSurface = widget.color ?? _colorScheme.onSurface;
    late Color dragColor;
    late Color hoverColor;
    late Color idleColor;
    dragColor = onSurface.withOpacity(0.9);
    hoverColor = onSurface.withOpacity(0.75);
    idleColor = onSurface.withOpacity(0.5);

    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
      if (states.contains(MaterialState.dragged))
        return _scrollbarTheme.thumbColor?.resolve(states) ?? dragColor;

      // If the track is visible, the thumb color hover animation is ignored and
      // changes immediately.
      if (states.contains(MaterialState.hovered) && _showTrackOnHover)
        return _scrollbarTheme.thumbColor?.resolve(states) ?? hoverColor;

      return Color.lerp(
        _scrollbarTheme.thumbColor?.resolve(states) ?? idleColor,
        _scrollbarTheme.thumbColor?.resolve(states) ?? hoverColor,
        _hoverAnimationController.value,
      )!;
    });
  }

  MaterialStateProperty<Color> get _trackColor {
    final Color onSurface = widget.color ?? _colorScheme.onSurface;

    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
      if (states.contains(MaterialState.hovered) && _showTrackOnHover) {
        return _scrollbarTheme.trackColor?.resolve(states) ??
            onSurface.withOpacity(0.05);
      }
      return const Color(0x00000000);
    });
  }

  MaterialStateProperty<Color> get _trackBorderColor {
    final Color onSurface = widget.color ?? _colorScheme.onSurface;

    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
      if (states.contains(MaterialState.hovered) && _showTrackOnHover) {
        return _scrollbarTheme.trackBorderColor?.resolve(states) ??
            onSurface.withOpacity(0.1);
      }
      return const Color(0x00000000);
    });
  }

  MaterialStateProperty<double> get _thickness {
    return MaterialStateProperty.resolveWith((Set<MaterialState> states) {
      if (states.contains(MaterialState.hovered) && _showTrackOnHover)
        return widget.hoverThickness ??
            _scrollbarTheme.thickness?.resolve(states) ??
            _kScrollbarThicknessWithTrack;
      // The default VsScrollbar thickness is smaller on mobile.
      return widget.thickness ??
          _scrollbarTheme.thickness?.resolve(states) ??
          (_kScrollbarThickness / (_useAndroidScrollbar ? 2 : 1));
    });
  }

  @override
  void initState() {
    super.initState();
    _hoverAnimationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 200),
    );
    _hoverAnimationController.addListener(() {
      updateScrollbarPainter();
    });
  }

  @override
  void didChangeDependencies() {
    final ThemeData theme = Theme.of(context);
    _colorScheme = theme.colorScheme;
    _scrollbarTheme = theme.scrollbarTheme;
    switch (theme.platform) {
      case TargetPlatform.android:
        _useAndroidScrollbar = true;
        break;
      case TargetPlatform.iOS:
      case TargetPlatform.linux:
      case TargetPlatform.fuchsia:
      case TargetPlatform.macOS:
      case TargetPlatform.windows:
        _useAndroidScrollbar = false;
        break;
    }
    super.didChangeDependencies();
  }

  @override
  void updateScrollbarPainter() {
    scrollbarPainter
      ..color = _thumbColor.resolve(_states)
      ..trackColor = _trackColor.resolve(_states)
      ..trackBorderColor = _trackBorderColor.resolve(_states)
      ..textDirection = Directionality.of(context)
      ..thickness = _thickness.resolve(_states)
      ..radius = widget.radius ??
          _scrollbarTheme.radius ??
          (_useAndroidScrollbar ? null : _kScrollbarRadius)
      ..crossAxisMargin = _scrollbarTheme.crossAxisMargin ??
          (_useAndroidScrollbar ? 0.0 : _kScrollbarMargin)
      ..mainAxisMargin = _scrollbarTheme.mainAxisMargin ?? 0.0
      ..minLength = _scrollbarTheme.minThumbLength ?? _kScrollbarMinLength
      ..padding = MediaQuery.of(context).padding;
  }

  @override
  void handleThumbPressStart(Offset localPosition) {
    super.handleThumbPressStart(localPosition);
    setState(() {
      _dragIsActive = true;
    });
  }

  @override
  void handleThumbPressEnd(Offset localPosition, Velocity velocity) {
    super.handleThumbPressEnd(localPosition, velocity);
    setState(() {
      _dragIsActive = false;
    });
  }

  @override
  void handleHover(PointerHoverEvent event) {
    super.handleHover(event);
    // Check if the position of the pointer falls over the painted VsScrollbar
    if (isPointerOverScrollbar(event.position, event.kind)) {
      // Pointer is hovering over the VsScrollbar
      setState(() {
        _hoverIsActive = true;
      });
      _hoverAnimationController.forward();
    } else if (_hoverIsActive) {
      // Pointer was, but is no longer over painted VsScrollbar.
      setState(() {
        _hoverIsActive = false;
      });
      _hoverAnimationController.reverse();
    }
  }

  @override
  void handleHoverExit(PointerExitEvent event) {
    super.handleHoverExit(event);
    setState(() {
      _hoverIsActive = false;
    });
    _hoverAnimationController.reverse();
  }

  @override
  void dispose() {
    _hoverAnimationController.dispose();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:vs_scrollbar/vs_scrollbar.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter VsScrollBar Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool _verticalList = true;
  ScrollController _scrollController = ScrollController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          backgroundColor: Colors.purple[900],
          title: Text("'Flutter vsScrollBar Demo'"),
        ),
        body: ListView(
          padding: EdgeInsets.all(20),
          physics: BouncingScrollPhysics(),
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: [
                ElevatedButton(
                    onPressed: () => setState(() => _verticalList = true),
                    child: Text("Vertical List")),
                ElevatedButton(
                    onPressed: () => setState(() => _verticalList = false),
                    child: Text("Horizontal List")),
              ],
            ),
            SizedBox(height: 20),
            Container(
              height: _verticalList
                  ? MediaQuery.of(context).size.height - 250
                  : 150,
              child: VsScrollbar(
                controller: _scrollController,
                showTrackOnHover: true, // default false
                isAlwaysShown: false, // default false
                scrollbarFadeDuration: Duration(
                    milliseconds: 500), // default : Duration(milliseconds: 300)
                scrollbarTimeToFade: Duration(
                    milliseconds: 800), // default : Duration(milliseconds: 600)
                style: VsScrollbarStyle(
                  hoverThickness: 10.0, // default 12.0
                  radius: Radius.circular(10), // default Radius.circular(8.0)
                  thickness: 10.0, // default 8.0
                  color: Colors.purple.shade900, // default ColorScheme Theme
                ),
                child: ListView.builder(
                    controller:
                        _scrollController, // use same scrollController object to support drag functionality
                    shrinkWrap: true,
                    physics: BouncingScrollPhysics(),
                    scrollDirection:
                        _verticalList ? Axis.vertical : Axis.horizontal,
                    itemCount: 6,
                    itemBuilder: (BuildContext context, int index) {
                      return Container(
                          height: 100,
                          width: MediaQuery.of(context).size.width,
                          color: Colors.purple[900],
                          margin: EdgeInsets.symmetric(
                              horizontal: 25, vertical: 20));
                    }),
              ),
            )
          ],
        ));
  }
}
